scripts/detection/center_net/train_center_net.py

```diff
@@ -0,0 +1,305 @@
+        batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers)
+    return train_loader, val_loader
+
+def save_params(net, best_map, current_map, epoch, save_interval, prefix):
+    current_map = float(current_map)
+    if current_map > best_map[0]:
+        best_map[0] = current_map
+        net.save_parameters('{:s}_best.params'.format(prefix, epoch, current_map))
+        with open(prefix+'_best_map.log', 'a') as f:
+            f.write('{:04d}:\t{:.4f}\n'.format(epoch, current_map))
+    if save_interval and epoch % save_interval == 0:
+        net.save_parameters('{:s}_{:04d}_{:.4f}.params'.format(prefix, epoch, current_map))
+
+def validate(net, val_data, ctx, eval_metric, flip_test=False):
+    """Test on validation dataset."""
+    eval_metric.reset()
+    net.flip_test = flip_test
+    mx.nd.waitall()
+    net.hybridize()
+    for batch in val_data:
+        data = gluon.utils.split_and_load(batch[0], ctx_list=ctx, batch_axis=0, even_split=False)
+        label = gluon.utils.split_and_load(batch[1], ctx_list=ctx, batch_axis=0, even_split=False)
+        det_bboxes = []
+        det_ids = []
+        det_scores = []
+        gt_bboxes = []
+        gt_ids = []
+        gt_difficults = []
+        for x, y in zip(data, label):
+            # get prediction results
+            ids, scores, bboxes = net(x)
+            det_ids.append(ids)
+            det_scores.append(scores)
+            # clip to image size
+            det_bboxes.append(bboxes.clip(0, batch[0].shape[2]))
+            # split ground truths
+            gt_ids.append(y.slice_axis(axis=-1, begin=4, end=5))
+            gt_bboxes.append(y.slice_axis(axis=-1, begin=0, end=4))
+            gt_difficults.append(y.slice_axis(axis=-1, begin=5, end=6) if y.shape[-1] > 5 else None)
+
+        # update metric
+        eval_metric.update(det_bboxes, det_ids, det_scores, gt_bboxes, gt_ids, gt_difficults)
+    return eval_metric.get()
+
+def train(net, train_data, val_data, eval_metric, ctx, args):
+    """Training pipeline"""
+    net.collect_params().reset_ctx(ctx)
+    # lr decay policy
+    lr_decay = float(args.lr_decay)
+    lr_steps = sorted([int(ls) for ls in args.lr_decay_epoch.split(',') if ls.strip()])
+    lr_decay_epoch = [e - args.warmup_epochs for e in lr_steps]
+    num_batches = args.num_samples // args.batch_size
+    lr_scheduler = LRSequential([
+        LRScheduler('linear', base_lr=0, target_lr=args.lr,
+                    nepochs=args.warmup_epochs, iters_per_epoch=num_batches),
+        LRScheduler(args.lr_mode, base_lr=args.lr,
+                    nepochs=args.epochs - args.warmup_epochs,
+                    iters_per_epoch=num_batches,
+                    step_epoch=lr_decay_epoch,
+                    step_factor=args.lr_decay, power=2),
+    ])
+
+    for k, v in net.collect_params('.*bias').items():
+        v.wd_mult = 0.0
+    # trainer = gluon.Trainer(
```

Comment:
Please remove comments if unnecessary. And is adam better than sgd?

Reply:
similar, though I don't have fully compariable results in the latest version